fortitude/src/fix/mod.rs

```diff
@@ -0,0 +1,365 @@
+            if last_pos.is_some_and(|last_pos| last_pos >= first.start()) {
+                continue;
+            }
+        }
+
+        let mut applied_edits = Vec::with_capacity(fix.edits().len());
+        for edit in edits {
+            // Add all contents from `last_pos` to `fix.location`.
+            let slice = locator.slice(TextRange::new(last_pos.unwrap_or_default(), edit.start()));
+            output.push_str(slice);
+
+            // Add the start source marker for the patch.
+            source_map.push_start_marker(edit, output.text_len());
+
+            // Add the patch itself.
+            output.push_str(edit.content().unwrap_or_default());
+
+            // Add the end source marker for the added patch.
+            source_map.push_end_marker(edit, output.text_len());
+
+            // Track that the edit was applied.
+            last_pos = Some(edit.end());
+            applied_edits.push(edit);
+        }
+
+        applied.extend(applied_edits.drain(..));
+        *fixed.entry(rule).or_default() += 1;
+    }
+
+    // Add the remaining content.
+    let slice = locator.after(last_pos.unwrap_or_default());
+    output.push_str(slice);
+
+    // TODO: Should be able to reuse locator here? Or don't store a
+    // SourceFile in FixResult, store something else?
+    let source_file = SourceFileBuilder::new(name, output).finish();
+
+    FixResult {
+        code: source_file,
+        fixes: fixed,
+        source_map,
+    }
+}
+
+/// Compare two fixes.
+fn cmp_fix(_rule1: Rule, _rule2: Rule, fix1: &Fix, fix2: &Fix) -> std::cmp::Ordering {
+    // Apply fixes in order of their start position.
+    fix1.min_start().cmp(&fix2.min_start())
+}
+
+#[cfg(test)]
+mod tests {
+    use ruff_diagnostics::{Diagnostic, Edit, Fix, SourceMarker};
+    use ruff_source_file::Locator;
+    use ruff_text_size::{Ranged, TextSize};
+
+    use crate::fix::{apply_fixes, FixResult};
+    use crate::rules::modules::use_statements::UseAll;
+
+    #[allow(deprecated)]
+    fn create_diagnostics(edit: impl IntoIterator<Item = Edit>) -> Vec<Diagnostic> {
+        edit.into_iter()
+            .map(|edit| Diagnostic {
+                // The choice of rule here is arbitrary.
+                kind: UseAll {}.into(),
+                range: edit.range(),
+                fix: Some(Fix::safe_edit(edit)),
+                parent: None,
+            })
+            .collect()
+    }
+
+    #[test]
+    fn empty_file() {
+        let locator = Locator::new(r"");
+        let diagnostics = create_diagnostics([]);
+        let FixResult {
+            code,
+            fixes,
+            source_map,
+        } = apply_fixes(diagnostics.iter(), &locator, "test.f90");
+        assert_eq!(code.source_text(), "");
+        assert_eq!(fixes.values().sum::<usize>(), 0);
+        assert!(source_map.markers().is_empty());
+    }
+
+    #[test]
+    fn apply_one_insertion() {
+        let locator = Locator::new(
+            r#"
+import os
+
+print("hello world")
```

Comment:
Could the tests be converted to something similar in Fortran?

Reply:
Yes, but these are actually all language agnostic, just working on `String`s. The magic numbers would need updating too, so I'm not sure it's worth the effort to be honest!